engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java

```diff
@@ -0,0 +1,105 @@
+package org.enso.interpreter.test;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.file.Paths;
+import java.util.Map;
+import org.enso.polyglot.RuntimeOptions;
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.Language;
+import org.graalvm.polyglot.Source;
+import org.graalvm.polyglot.Value;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PolyglotErrorTest {
+  private Context ctx;
+  private Value panic;
+
+  public static String bar(Object o) {
+    return "[[" + o + "]]";
```

Comment:
If one wants to control the exception behavior then I am afraid one has to use `Value` - for example:
```diff
--- a/engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java
+++ b/engine/runtime/src/test/java/org/enso/interpreter/test/PolyglotErrorTest.java
@@ -18,9 +18,22 @@
   private Context ctx;
   private Value panic;
 
-  public static String bar(Object o) {
-    return "[[" + o + "]]";
+  public static String bar(Value v) throws Exception {
+    String s;
+    if (v.isString()) {
+      s = v.asString();
+    } else {
+      var t = v.invokeMember("to_text");
+      if (t.isString()) {
+        s = t.asString();
+      } else if (t.isException()) {
+        throw t.as(Exception.class);
+      } else {
+        s = t.toString();
   }
+    }
+    return "[[" + s + "]]";
+  }
 
   @Before
   public void prepareCtx() throws Exception {
```
allows one to observe `Panic` (via `catch PolyglotException`) and `Error` (via `isException()`).

Reply:
@radeusgd, regardless of the [outcome of the GraalVM slack discussion](https://graalvm.slack.com/archives/CNQSB2DHD/p1673436829788349?thread_ts=1673412770.137559&cid=CNQSB2DHD), the previous snippet gives you a way to control, propagate and handle exceptions from `to_text` if necessary. Works in current GraalVM version.

Reply:
Oh, that is a pretty sensible workaround!n Thanks for showing it and also I somehow missed it.